### ISSUE TITLE
When stream api send a notification?

### ISSUE THREAD

**Issue by mattdamon108** [GetStream/stream-chat-react-native#72]
I implemented APNs and FCM for push notification for my react native app. It works fine and gets notifications in the foreground/background/when the app is closed. I tested it with APNs notification send utility and stream CLI push:test.

But, in the case of iOS, It seems the push notification is sent when the app is closed. When I test with stream CLI, it gets a push notification when the app is in the background too. I guess stream API starts sending a notification when a channel member seems offline or something.

How can I make my app getting a notification from stream API even though my app is in foreground or background?

**Comment by vishalnarkhede** [comment]
Hey @mattdamon108 ,

Currently we send push notification only when ws connection is NOT active for a particular user/device. We have plans to improve this behavior in our roadmap, so that you will be able to handle it on client side. I will keep you posted.

**Comment by mattdamon108** [comment]
Hi @vishalnarkhede ,

Well noted. I'm happy to hear your plan. But can I get any recommendations to connect/disconnect ws to handle the trigger for a notification? As per the doc, I found `client.disconnect()`, but it seems heavy to use it while my app is running and between screens. (I'd like to handle the notification trigger while my app is even foreground.)

**Comment by rgkobashi** [comment]
Probably the author of this has solved the issue already, but for those who haven't (this behavior caught me for surprise too) what I did was creating our own local notification:

```
Notifications.shared.localNotificationContent = { (message, channel) in
    // parse message/channel and create/return your own UNNotificationContent
}
```

Notice that this code is for the iOS client. But the important thing to notice is that, when the app is sent to background, Stream listens for new messages through a web socket, until the OS stops (not kill) the app (iOS apps can run code after being sent to background for a period of time defined by the OS). Between these time, Stream makes a local notification (looks like a push notification, but it is not).

Fortunately Stream provides a closure (the snipped above) so we can customize this local notification:

**Comment by BrantApps** [comment]
Yes, this is how we worked around it at Candide too ... certainly not a fix to this issue - the library needs to listen for AppState changes and behave as the user would expect and as the developer would expect.

Lots of time has been wasted by paying customers of this library wondering why it can't do chat behaviours properly. 🙄

**Comment by lexin** [comment]
Can I ask for your help? I tried to setup local notifications just like it is mentioned 
_Notifications.shared.localNotificationContent = { (message, channel) in
            let content = UNMutableNotificationContent()
                  content.title = "\(message.user.name) @ \(channel.name ?? "")"
                  content.body = message.text
                  content.sound = UNNotificationSound.default
                 content.badge = (UIApplication.shared.applicationIconBadgeNumber + 1) as NSNumber
                 return content
        }_
But it doesn't work - if the app in the background no notifications. 
But if I close it completely, Push notifications work properly. (User granted access for push notifications)

**Comment by rgkobashi** [comment]
Are you watching the channels? In my app `ChannelsViewController` is on a tab and I didnt get those notifications (local ones) until I switched to that tab. 

The thing is, you will get local notifications after you watch the channels, this means after you show `ChannelsViewController ` which Stream starts  watching them automatically for you, or you start watching them manually.

Took me a while to figure this out, documentation is not very clear :(

**Comment by lexin** [comment]
Thank you for your response. Yes, I am watching channels. 
And I tried to turn on watching in two different ways: 
1 add option .watch to query 
2 call watch for each channel manually. 

 is it important when and where to call the code 
_Notifications.shared.localNotificationContent = { (message, channel) in
            let content = UNMutableNotificationContent()_

I do it just after a successful Client init. 

I can manage local notifications by myself - just to present it when catching the notificationMessageNew or messageNew in the subscription but it's strange why  native tool doesn't work

**Comment by rgkobashi** [comment]
hmmm... 🤔
Are you able to see the code of the SDK, what I did to realized about this behavior (and a lot more) was setting breakpoints on the SDK code, you can check `Notifications` source code, `Notifications.showIfNeeded(message:channel:)` is a good start point.

Oh, one more thing, somewhere in your code are you conforming to `UNUserNotificationCenterDelegate`? Since `UNUserNotificationCenter` is a singleton, Stream makes use of that one, so if you are using it too, Stream wont be able to use it, thus no local notifications.

> is it important when and where to call the code

I am calling `Notifications.shared.xxx` after this:

```
let config = Client.Config(...)
Client.configureShared(config)
```

**Comment by lexin** [comment]
Unfortunately, I don't see the code itself. I just installed the GetStream via cocoapod and get the framework only. 
Probably, it's to ask tech support from getstream team. Thank you.

**Comment by rgkobashi** [comment]
I also installed it with cocoapods and I can see the code, it has been very useful, it is well documented, but yeah, it will probably be better to contact support 👍

<img width="293" alt="Screen Shot 2020-11-27 at 7 01 50 PM" src="https://user-images.githubusercontent.com/4319094/100436768-24eaa900-30e3-11eb-8499-7a48ac19736b.png">

**Comment by lexin** [comment]
Just made an own solution 
 _let subscriptionEvent = Client.shared.subscribe { (event) in
            if ((event.type == .messageNew) || (event.type == .notificationMessageNew)) {
                switch event {
                case .messageNew(let message, let unreadCount, _, let channelID, _):
                    let channelname = self.getLocalChannel(id: channelID?.id ?? "123" )?.channel.name ?? ""
                    self.showLocalNotification(...)

                case .notificationMessageNew(let message, let channel, let unreadCount, _, _):
               	  self.showLocalNotification(...)

                default:
                    print ("")
                }

            }_

**Comment by sergeyzhukov** [comment]
@vishalnarkhede any updates on 

**Members that are currently online do not receive push messages** 

?

Our service needs to send push notifications for users across multiple devices. And if a user is active on the web (but not in the channel where the new message received) - a push notification should be sent to the iOS device.

**Comment by ferhatelmas** [comment]
@sergeyzhukov this is the behavior of v1 of the push. v2 which is in the process of being live at the end of this month, has app and user level configurations. One of them is to be able to send irrespective online status, by setting that flag, apps can retrieve notifications even if they are online in other devices. For the online device, client should ignore it though.

**Comment by rgkobashi** [comment]
Hi @ferhatelmas thanks for the response.
One more question, are we going to get notified about this once it is available? Or where does this flag needs to be enable, on the server side or on Stream dashboard?
TIA

**Comment by sergeyzhukov** [comment]
@ferhatelmas @vishalnarkhede Any updates related to push notifications delivery logic?

**Comment by ferhatelmas** [comment]
@sergeyzhukov More or less next week, docs should be updated on how to enable v2, update general app config and send user preferences for further customization.
